server/src/main/java/io/spine/server/delivery/ShardedStream.java

```diff
@@ -0,0 +1,361 @@
+                                       .getClassName();
+        final ShardIndex shardIndex = key.getIndex();
+
+        final String value = on("__").join("bc_", boundedContextName.getValue(),
+                                           "target_", className,
+                                           "prd_", Stringifiers.toString(shardIndex),
+                                           "env_", envelopeClass);
+        final StringValue asMsg = StringValue.newBuilder()
+                                             .setValue(value)
+                                             .build();
+        final Any asAny = AnyPacker.pack(asMsg);
+        final ChannelId result = ChannelId.newBuilder()
+                                          .setIdentifier(asAny)
+                                          .build();
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        final ShardedStream<?, ?, ?> that = (ShardedStream<?, ?, ?>) o;
+        return Objects.equals(boundedContextName, that.boundedContextName) &&
+                Objects.equals(key, that.key) &&
+                Objects.equals(tag, that.tag);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(boundedContextName, key, tag);
+    }
+
+    @Override
+    public String toString() {
+        return MoreObjects.toStringHelper(this)
+                          .add("Bounded Context name", boundedContextName)
+                          .add("Sharding key", key)
+                          .add("Delivery tag", tag)
+                          .add("Target ID class", targetIdClass)
+                          .toString();
+    }
+
+    /**
+     * The observer of stream messages.
+     *
+     * <p>Passes the received messages to the respective {@linkplain ShardedStreamConsumer
+     * delegate}.
+     */
+    private class ExternalMessageObserver implements StreamObserver<ExternalMessage> {
+
+        private final ShardedStreamConsumer<I, E> delegate;
+
+        private ExternalMessageObserver(ShardedStreamConsumer<I, E> delegate) {
+            this.delegate = delegate;
+        }
+
+        @Override
+        public void onNext(ExternalMessage value) {
+            checkNotNull(value);
+            final ShardedMessage shardedMessage = ExternalMessages.asShardedMessage(value);
+
+            final E envelope = converter().envelopeOf(shardedMessage);
+            final I targetId = converter().targetIdOf(shardedMessage, getTargetIdClass());
+            checkNotNull(envelope);
+            delegate.onNext(targetId, envelope);
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            throw newIllegalStateException(t,
+                                           "Error observing the external messages for consumer " +
```

Comment:
Consider having the formatting parameter on a new line, which would allow to keep the string one piece.